lib/permissions.js

```diff
@@ -0,0 +1,573 @@
+              // if any existing, retrieve the agents having this auth rule
+              // assigned
+              if (readStmts && readStmts.length > 0) {
+                for (i in readStmts) {
+                  var stmt = readStmts[i]
+                  var agent = getAgent(graph, stmt)
+                  addRule(permissions, agent.uri, PermissionType.Read)
+                }
+              }
+
+              // read all the statements having Write auth rule
+              var writeStmts = graph.statementsMatching(undefined, vocab.acl('mode'), vocab.acl('Write'))
+              // if any existing, retrieve the agents having this auth rule
+              // assigned
+              if (writeStmts && writeStmts.length > 0) {
+                for (i in writeStmts) {
+                  stmt = writeStmts[i]
+                  agent = getAgent(graph, stmt)
+                  addRule(permissions, agent.uri, PermissionType.Write)
+                }
+              }
+
+              // read all the statements having Append auth rule
+              var appendStmts = graph.statementsMatching(undefined, vocab.acl('mode'), vocab.acl('Append'))
+              // if any existing, retrieve the agents having this auth rule
+              // assigned
+              if (appendStmts && appendStmts.length > 0) {
+                for (i in appendStmts) {
+                  stmt = appendStmts[i]
+                  agent = getAgent(graph, stmt)
+                  addRule(permissions, agent.uri, PermissionType.Append)
+                }
+              }
+
+              // clean duplicates
+              for (var key in permissions) {
+                var agents = permissions[key]
+                permissions[key] = removeArrayDuplicates(agents)
+              }
+
+              // successfully retrieved the resource permissions. return the
+              // permissions object
+              resolve(permissions)
+            }
+          ).catch(
+            function (err) {
+              reject(err)
+            }
+          )
+        }
+      ).catch(
+        function (err) {
+          reject(err)
+        }
+      )
+    })
+  },
+  /**
+   * Given a resource url, retrieve the url of the acl resource that has the
+   * permission rules for this resource. Note that the retrieved url may be the
+   * url of on ancestor of the resource.
+   * @method getAclResource
+   * @param url {String} The url of the resource to retrieve the acl resource
+   * url for
+   * @return {Promise<String>} The url of the acl resource
+   */
+  getAclResource: function getAclResource (url) {
+    var self = this
+
+    return new Promise(function (resolve, reject) {
+      webClient.head(url).then(
+        function (solidResponse) {
+          var aclUrl
+          // if the resource exists and there is a logged in user, then
+          // succeed with the acl url
+          if (solidResponse.exists() && solidResponse.isLoggedIn()) {
```

Comment:
No need to check if the user is logged in - a `web.head()` request returns an acl link regardless of logged-in status.